crates/rpc/src/dto/receipt.rs

```diff
@@ -164,7 +163,11 @@ impl SerializeForVersion for TxnReceiptWithBlockInfo<'_> {
         })?;
 
         serializer.serialize_optional("block_hash", block_hash.cloned())?;
-        serializer.serialize_optional("block_number", *block_number)?;
+        if (serializer.version == RpcVersion::V09)
+            || (serializer.version != RpcVersion::V09 && block_hash.is_some())
+        {
+            serializer.serialize_optional("block_number", *block_number)?;
+        }
```

Comment:
I'd rather use a condition here that won't break when we're adding the next JSON-RPC specification.

I think something like this would be slightly more future-proof:

```suggestion
        if (serializer.version >= RpcVersion::V09)
            || (serializer.version < RpcVersion::V09 && block_hash.is_some())
        {
            serializer.serialize_optional("block_number", *block_number)?;
        }
```